pkg/sanity/backup_restore.go

```diff
@@ -0,0 +1,423 @@
+					SrcVolumeID: volumeID,
+				}
+				bkpStatusResp = volumedriver.BackupStatus(bkpStatusReq)
+				Expect(bkpStatusResp.StsErr).To(BeNil())
+
+				bkpStatus = bkpStatusResp.Statuses[volumeID]
+				if strings.Contains(bkpStatus.Status, "Done") {
+					break
+				}
+				if strings.Contains(bkpStatus.Status, "Active") {
+					time.Sleep(time.Second * 10)
+				}
+				if strings.Contains(bkpStatus.Status, "failed") {
+					// give 3 attempts for backup to be successfull before declaring failed
+					time.Sleep(time.Second * 10)
+					retries++
+				}
+			}
+
+			Expect(bkpStatus.Status).To(BeEquivalentTo("Done"))
+
+			By("Backup enumerate")
+
+			bkpEnumReq := &api.BackupEnumerateRequest{
+				BackupGenericRequest: api.BackupGenericRequest{
+					All:            false,
+					CredentialUUID: credUUID,
+					SrcVolumeID:    volumeID,
+				},
+			}
+
+			enumResp := volumedriver.BackupEnumerate(bkpEnumReq)
+			Expect(enumResp.EnumerateErr).To(BeNil())
+			Expect(len(enumResp.Backups)).ToNot(BeNil())
+			Expect(enumResp.Backups[0].SrcVolumeID).To(BeEquivalentTo(volumeID))
+		})
+	})
+
+	Describe("Volume Backup Restore", func() {
+
+		var (
+			numVolumesBefore int
+			numVolumesAfter  int
+			volumeID         string
+			bkpStatusReq     *api.BackupStsRequest
+			bkpStatusResp    *api.BackupStsResponse
+			bkpStatus        api.BackupStatus
+			restoredVolume   string
+		)
+
+		BeforeEach(func() {
+			var err error
+			restoredVolume = "backup-restored"
+			volumes, err := volumedriver.Enumerate(&api.VolumeLocator{}, make(map[string]string))
+			numVolumesBefore = len(volumes)
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		AfterEach(func() {
+			var err error
+
+			err = volumedriver.Delete(volumeID)
+			Expect(err).ToNot(HaveOccurred())
+
+			err = volumedriver.Delete(restoredVolume)
+			Expect(err).ToNot(HaveOccurred())
+
+			volumes, err := volumedriver.Enumerate(&api.VolumeLocator{}, nil)
+			Expect(err).ToNot(HaveOccurred())
+			numVolumesAfter = len(volumes)
+		})
+
+		It("Should restore backup", func() {
+
+			By("Creating the volume")
+
+			var err error
+			var size = 1
+			vr := &api.VolumeCreateRequest{
+				Locator: &api.VolumeLocator{
+					Name: "create-to-restore-backup",
+					VolumeLabels: map[string]string{
+						"class": "restore-class",
+					},
+				},
+				Source: &api.Source{},
+				Spec: &api.VolumeSpec{
+					Size:      uint64(size * GIGABYTE),
+					HaLevel:   1,
+					Format:    api.FSType_FS_TYPE_EXT4,
+					BlockSize: 128,
+					Cos:       api.CosType_LOW,
+				},
+			}
+
+			volumeID, err = volumedriver.Create(vr.GetLocator(), vr.GetSource(), vr.GetSpec())
+
+			By("Checking if volume created successfully with the provided params")
+			testIfVolumeCreatedSuccessfully(volumedriver, volumeID, numVolumesBefore, vr)
+
+			By("Doing Backup")
+
+			bkpReq := &api.BackupRequest{
+				CredentialUUID: credUUID,
+				Full:           false,
+				VolumeID:       volumeID,
+			}
+
+			time.Sleep(time.Second * 10)
```

Comment:
Why sleep? Open ended feedback is erroneous. Try to find another model for detecting a status or event change.

Reply:
done